_posts/2016-06-08-component-relative-paths-in-angular-2.md

```diff
@@ -0,0 +1,305 @@
+*  What if my components are organized in distinct packages ?
+*  What if we want to organize our components by feature (Angular Style Guide - Best Practice)?
+
+To explore the issue, let's consider the scenario where our details component (and files) are in the `src/app/header` package. The urls used above would cause the loader to fail and the developer would see the following 404 error in the developers console:
+
+![component_url_404](https://cloud.githubusercontent.com/assets/210413/15878482/eca5cba2-2ce0-11e6-8fb4-78868bec2644.png)
+
+Are you tempted to try and debug or introspect the exception stack and determine why the file was not found: good luck with that!
+ 
+So if path to the component HTML or CSS file is not valid, the **EASY** workaround is to add absolute paths to the URLs... but don't do it:
+
+
+**header.component.ts**
+{% highlight ts %}
+{% raw %}
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector   : 'contacts-header',
+  templateUrl: 'src/app/header/header.component.html',
+  stylesUrl  : 'src/app/header/header.component.css'
+})
+export class HeaderComponent implements OnInit {
+  constructor() {}
+  ngOnInit() {}
+}
+{% endraw %}
+{% endhighlight %}
+
+I hope you agree that this is a horrible idea and band-aid solution. 
+
+* What if I move my components to other packages? I will have to update the absolute paths...
+* What if I want to reuse my components in other applications?
+* Can I not use using relative-paths in my components ?
+
+![what](https://cloud.githubusercontent.com/assets/210413/15881627/0d0bc056-2cfd-11e6-8086-a87bfbbf91e6.gif)
+
+
+## Components with Relative-Path URLs
+
+In fact, you can use relative paths. But not the way you may first think... and not without understanding and accepting some constraints.
+
+At first we might be tempted to try this:
+
+**header.component.ts**
+{% highlight ts %}
+{% raw %}
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector   : 'contacts-header',
+  templateUrl: './header.component.html',
+  stylesUrl  : './header.component.css'
+})
+export class HeaderComponent implements OnInit {
+  constructor() {}
+  ngOnInit() {}
+}
+{% endraw %}
+{% endhighlight %}
+
+You might expect that `./header.component.html` is a path relative to the `header.component.ts` file and that component-relative paths should work, right?  Nope, you get another **404 Not found Exception**.
+
+Remember we noted that the paths are relative to the Application Root **at load time**? Since we are using the package `src/app/header/header.component.*`, then our files obviously are not at the app root.
+
+We could use a gulp or grunt task to deploy to a `dist` directory and have all our components in the dist root directory. 
+
+OMG, that is a horrible idea! Don't do it.
+
+![no-gif](https://cloud.githubusercontent.com/assets/210413/15881568/9c552dac-2cfc-11e6-808c-f84540d2d758.gif)
```

Comment:
I would like to keep this one... it adds some light comedy to emphasize our message.